cpp/src/arrow/c/bridge.cc

```diff
@@ -1777,17 +1779,30 @@ class ArrayStreamBatchReader : public RecordBatchReader {
     return Status::OK();
   }
 
- private:
-  std::shared_ptr<Schema> CacheSchema() const {
-    if (!schema_) {
-      struct ArrowSchema c_schema;
-      ARROW_CHECK_OK(StatusFromCError(stream_.get_schema(&stream_, &c_schema)));
-      schema_ = ImportSchema(&c_schema).ValueOrDie();
+  static Result<std::shared_ptr<RecordBatchReader>> Make(
+      struct ArrowArrayStream* stream) {
+    if (ArrowArrayStreamIsReleased(stream)) {
+      return Status::Invalid("Cannot import released ArrowArrayStream");
+    }
+    std::shared_ptr<Schema> schema;
+    struct ArrowSchema c_schema = {};
+    Status status = StatusFromCError(stream, stream->get_schema(stream, &c_schema));
+    if (status.ok()) {
+      status = ImportSchema(&c_schema).Value(&schema);
     }
-    return schema_;
+    if (!status.ok()) {
+      stream->release(stream);
```

Comment:
Should we use `ArrowArrayStreamRelease()` here?

```suggestion
      ArrowArrayStreamRelease(stream);
```